### ISSUE TITLE
TensorForestEstimator with input_fn results in infinite loop for evaluate and predict.

### ISSUE THREAD

**Issue by martarek** [tensorflow/tensorflow#5834]
When using `TensorForestEstimator`, using the previous implementation with x= and y= works. However, when trying to convert to the soon-to-be standard of input_fn, I get strange behaviors. Given 

```
params = tf.contrib.tensor_forest.python.tensor_forest.ForestHParams(
num_trees=10,
max_nodes=100,
num_classes=2,
num_features=features_count)
classifier = TensorForestEstimator(params)
```

calling 
```
classifier.fit(input_fn=(lambda: training_input_fn(trainset)))
```
works and completes correctly. The input_fn used is the following :

```
features_name = [] # ordered list of all features in the dataset
def training_input_fn(dataset):
        data, target = dataset
        # 'data' is a numpy array where data[:,i] returns all observations for a given feature
        # It is reshaped because otherwise I had a concat error
        features = {features_name[i]: tf.constant(data[:, i], shape=(data.shape[0], 1))
                    for i in range(len(features_name)) if features_name[i] in DEFAULT_FEATURES}
        labels = tf.constant(target)
        return features, labels
```

However, calling 
```
classifier.evaluate(input_fn=(lambda: training_input_fn(validset)))
```
afterwards results in an apparently neverending loop in `tf.contrib.learn.python.learn.graph_actions` in function `run_feeds_iter` in the following snippet :
```
      try:
        threads = queue_runner.start_queue_runners(session, coord=coord)
        for f in feed_dicts:
          yield session.run(output_dict, f)
      finally:
        coord.request_stop()
        if threads:
          coord.join(threads, stop_grace_period_secs=120)
```
The loop seems like never stopping. Also, memory starts increasing slowly but steadily from there (had to stop at 10 GB for this fairly small dataset). The same goes if I try predict on it. 
My Training dataset is composed of 3633 entries with 183 features and my validation dataset is composed of 2180 entries. 

### Environment info
Operating System: 
Ubuntu 16.04, running on Python3.5

Installed version of CUDA and cuDNN: 
```
-rw-r--r-- 1 root root   558720 Nov 22 10:52 /usr/local/cuda/lib64/libcudadevrt.a
lrwxrwxrwx 1 root root       16 Nov 22 10:52 /usr/local/cuda/lib64/libcudart.so -> libcudart.so.8.0
lrwxrwxrwx 1 root root       19 Nov 22 10:52 /usr/local/cuda/lib64/libcudart.so.8.0 -> libcudart.so.8.0.44
-rwxr-xr-x 1 root root   415432 Nov 22 10:52 /usr/local/cuda/lib64/libcudart.so.8.0.44
-rw-r--r-- 1 root root   775162 Nov 22 10:52 /usr/local/cuda/lib64/libcudart_static.a
lrwxrwxrwx 1 root root       13 Nov 22 12:47 /usr/local/cuda/lib64/libcudnn.so -> libcudnn.so.5
lrwxrwxrwx 1 root root       17 Nov 22 12:47 /usr/local/cuda/lib64/libcudnn.so.5 -> libcudnn.so.5.1.5
-rwxr-xr-x 1 root root 79337624 Nov 22 12:47 /usr/local/cuda/lib64/libcudnn.so.5.1.5
-rw-r--r-- 1 root root 69756172 Nov 22 12:47 /usr/local/cuda/lib64/libcudnn_static.a
```

If installed from source, provide 
1. The commit hash (`git rev-parse HEAD`) : `a26a5925b0500d0e9cd259989fc0e113fa29e27f`
2. The output of `bazel version` : 

`Build label: 0.4.0
Build target: bazel-out/local-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Wed Nov 2 17:54:14 2016 (1478109254)
Build timestamp: 1478109254
Build timestamp as int: 1478109254`

**Comment by prb12** [comment]
@gilberthendry Is this the inteneded use of `input_fn`? - I can't tell from the documentation!

I don't know anything about `TensorForestEstimator`, but in other places where an `input_fn` is used in TensorFlow it has often been to specify code to read a mini-batch from the file system using some sort of IO op, preprocess the samples and insert into a `Queue` of some kind.  In these other scenarios, the sub-graph created by `input_fn` is executed repeatedly in a `QueueRunner` thread.

In your example, you seem to be reading the entire dataset into a `tf.constant` and adding it to the graph.  This is probably fine if input_fn is only called once.

If `input_fn` was called repeatedly, then you would a) have a huge memory leak by adding large constants to the graph, and b) get dreadful performance by repeatedly changing the graphdef causing all the cached optimized datastructures to be thrown away.

You could probably check this quickly by adding a print/log statement to `training_input_fn`.

**Comment by gilberthendry** [comment]
As @prb12 points out, this isn't the usual usage of an input function (returning a tf.constant).  For starters, you won't get any mini-batching.  It looks like you'd be processing the whole dataset on every step during training (which is ok if that's what you want, but is a not very efficient thing to do for online extremely random forests). 

Without trying to reproduce this myself, I suspect that evaluate might never terminate because it never throws the exception that indicates that the input is out of data (I think it still works this way).  You can verify this by passing max_steps=1 to evaluate.  

Still, it seems strange that the input_fn is being called more than once, resulting in your memory leak and such.  

I don't know what the migration plan is for other estimators to switch to input_fn, but TensorForestEstimator will continue to support x= and y= for datasets that easily fit in memory.  We will do the conversion to the right kind of tf.learn.Estimator under the hood (there will be an explicit class for this).  So my advice is to keep using x= and y= unless you want to use the IO ops mentioned by @prb12, which is really only necessary for large datasets.

**Comment by prb12** [comment]
> Still, it seems strange that the input_fn is being called more than once, resulting in your memory leak and such.

Note - that was just my hypothesis - we haven't yet verified whether this is what's happening.

**Comment by martarek** [comment]
@prb12 Thanks for the feedback. My `training_input_fn` is only called once for fit and once for evaluate. However, if I add an print statement with and increasing integer in  `run_feeds_iter` mentioned earlier, it just prints, _ad infinitum_, which leads me to believe that @gilberthendry might be right about the exception never being raised.

@gilberthendry Thanks for the clarification. I tried passing "max_steps=1" to evaluate, but the argument does not exist for evaluate. Also, I believe you may be right about the exception never being raised. What is the proper way of raising that exception in an `input_fn`? And the strangest, why does my training ( `fit` function) works while it fails on `predict` and `evaluate` ... They are using the same `input_fn`, with the exception of the data used being different. There is something I am missing here... 

On a side note, I was using x and y inputs before, but now I get a Warning message saying the x= and y= inputs will be deprecated and to use input_fn instead (or SKCompat). The resulting input_fn came from this [tutorial](https://www.tensorflow.org/versions/r0.11/tutorials/input_fn/index.html#building-the-input-fn). As I want to be able to export my trained model on Android, I want to use the [export](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/g3doc/api_docs/python/functions_and_classes/shard2/tf.contrib.learn.BaseEstimator.md#tfcontriblearnbaseestimatorexportargs-kwargs-baseestimatorexport) function to save my model. Since this function is not part of the SKCompat wrapper, I thought I had to convert my `x` and `y` to an `input_fn` to be compatible with future versions and with the `export` function itself which requires an `input_fn`. I still am struggling at understanding how I am suppose to pass an input to get a prediction out of the trained model once it is exported on Android, but I understand if this might be better posted on stackoverflow for support!

**Comment by gilberthendry** [comment]
My bad, the parameter is called just 'steps', not max_steps. Try that.

Fit works fine because TensorForestEstimator (like many estimators) doesn't rely on the input being exhausted to complete.  It uses TensorForestLossHook (or a custom hook) to terminate training when the forest stops growing.

The way to indicate that input is exhausted is to raise StopIteration.  The problem with your input_fn is that there's no place to raise StopIteration, because it isn't based on a get_feed_fn like data_feeder.py uses. 

You might be able to pass a separate input_fn for export (that maybe returns an appropriately typed and sized placeholder?), and use x=, y= for training and evaluation.  Not sure about that, just something to think about.


Overall, one way to go is to convert your data into tf.Examples, then using input_fn's is much more natural. Another way is to always evaluate with steps=1, and maybe exporting/predict will work fine.

**Comment by hughperkins** [comment]
I had exactly the same problem, with a custom estimator, and an `input_fn`, that returns a tensor, and just looping forever.

The fix turned out to be easy in the end, though whether it should really be necessary is an open question.  Instead of eg:

```
for y in estimator.predict(input_fn=...):
    print('y', y)
```
or:
```
y = list(estimator.predict(input_fn=...))
```
... use itertools.islice to only retrieve as many predictions as you inputed input data, ie/eg:
```
y_generator = estimator.predict(input_fn=...)
y = list(itertools.islice(y_generator, X.shape[0])
```

**Comment by martarek** [comment]
Thanks for all the feedback. I will consider the issue as closed for the moment because of all the proposed fixes that I have yet to test. Thanks for the quick support!

**Comment by sahildua2305** [comment]
I came here stuck at the same issue and the comment by @hughperkins solves it for me. Good catch! Thanks 😄

**Comment by jzhongaa** [comment]
I found that this issue happens in windows but when I run the exact same code in linux server this issue will not persist.
